contracts/src/BoundlessMarket.sol

```diff
@@ -359,42 +367,67 @@ contract BoundlessMarket is
+    {
         RequestId id = fill.id;
         (address client, uint32 idx) = id.clientAndIndex();
         Account storage clientAccount = accounts[client];
         (bool locked, bool fulfilled) = clientAccount.requestFlags(idx);
 
-        // Before other processing logic, check whether the request is marked as expired, from
-        // execution of the price request function. We do not emit the ProofDelivered event, and we
-        // do not issue a callback. This makes interpretation of the ProofDelivered events simpler,
-        // as they cannot be emitted for an expired request.
-        //
-        // Note that this will be true if and only if the request with the given digest is both
-        // valid (i.e. signed and internally consistent) and expired. If the request has not been
-        // priced (e.g. it is currently locked, and so a call to priceRequest was not required; or
-        // the request is invalid) then this bit will not be set. If request is not locked, and
-        // the valid bit is not set, the fulfillAndPayX functions will revert.
+        // Fetch the lock and fulfillment information.
+        // NOTE: The `lock` should only be used in code paths where locked is true.
+        RequestLock memory lock;
+        if (locked) {
+            lock = requestLocks[id];
+        }
         FulfillmentContext memory context = FulfillmentContextLibrary.load(fill.requestDigest);
-        if (context.expired) {
-            paymentError = abi.encodeWithSelector(RequestIsExpired.selector, RequestId.unwrap(id));
-            emit PaymentRequirementsFailed(paymentError);
-            return paymentError;
+
+        // First, check whether the request is known to be a valid signed request, and whether it is
+        // expired. If the request cannot be authenticated, revert.
+        //
+        // In the expired case, we return early here. We do not emit the ProofDelivered event, and
+        // we do not issue a callback. This makes interpretation of the ProofDelivered events
+        // simpler, as they cannot be emitted for an expired request.
+        if (context.valid) {
+            // Request has been validated in priceRequest, check the reported expiration.
+            if (context.expired) {
+                paymentError = abi.encodeWithSelector(RequestIsExpired.selector, RequestId.unwrap(id));
+                emit PaymentRequirementsFailed(paymentError);
+                return (paymentError, true);
+            }
+        } else if (locked && lock.requestDigest == fill.requestDigest) {
+            // Request was validated in lockRequest, check the expiration on the lock.
+            if (lock.deadline() < block.timestamp) {
+                paymentError = abi.encodeWithSelector(RequestIsExpired.selector, RequestId.unwrap(id));
+                emit PaymentRequirementsFailed(paymentError);
+                return (paymentError, true);
+            }
+        } else {
```

Comment:
One thing I'm thinking through is that it looks like its possible to price a request while its still locked, which would cause it to skip L400-406. It seems we are covered, but just wanted to flag, and if true might be worth adding a comment